### DISCUSSION TITLE
Replicator fieldtype

### DISCUSSION THREAD

**Discussion by kaidesu** [fusioncms/fusioncms#18] [Ideas]
The replicator fieldtype is our most advanced and most finicky fieldtype offered. While I would say the initial implementation of it was a success in v5 (I mean... it works), after much use I feel we've come to understand where it's initial limitations/difficulties lie. With v6 we're in a great place to rethink how data is stored and retrieved.

In a nutshell, the replicator fieldtype basically jams the entire matrix system into a single field. This allows content curators to develop "mini" matrices to store various sets of content - each with their own unique set of fields and configurations.

In v5, we stored this information as JSON across a variety of locations.

1. The initial configuration of the replicator field was stored with the field itself as json - which contained the various sets and their field configurations.
2. Then, when a new "entry" was created for any given set, we not only stored the _values_ of the fields, but also a carbon copy of the replicator configuration itself. This was required to bootstrap the replicator sets and fields without having to do some crazy deep database querying any time we wanted to access a replicator field.

Needless to say, there is *a lot* to manage there, and it gets messy pretty quickly. We can do better.

Instead of storing and jamming everything into a single JSON column, let's give our data and content some breathing room. Let's handle the replicator like we do the matrix, by dynamically generating the necessary database infrastructure to house the data. The database already handles data type conversion and storage for us, so that already is paying off as we've ran into many issues having to manually define and cast values based on their configured fieldtypes.

I think we can go about this in two ways:

1. Utilize the existing MySQL database and create a new table for each _set_ in a replicator fieldtype. This is exactly how the matrix operates, however, it can and will quickly bloat the database with a lot of tables if the replicator fieldtype is leaned upon for larger websites. While this won't lead to any performance hit, it'll just be a little chaotic when connecting to your website's database and being presented with 50+ tables. Ouch.
2. Create SQLite databases for each replicator/set, which we can store inside subdirectories to keep clean. Through the power of Laravel and some magic, we can configure dynamic database connections to be set up automatically whenever accessing any given replicator.

I'm leaning towards option 2 here. But we'll need to vet out any limitations/gotcha's to utilizing multiple, separate, and dynamic SQLite databases for storing content. One thing that comes to my immediate mind is search indexing.

**Comment by thekanbo** [comment]
I agree that connecting to and viewing a database with an excess number of tables is unappealing, and can make it hard to navigate the database if manual maintenance or retrieval is necessary.

That said, I don't think that relying on a secondary database engine _really_ helps that- storing those tables in a separate location doesn't solve the problem of making management easier, since accessing that data would have to be done outside of the DB browser that's being used to view the main database.

To me, it seems like both approaches are valid, but that they both have the same drawback of being difficult to manage by hand. To that end, storing the replicator data outside of the MySQL database is more cumbersome because that data doesn't exist in the same place as the rest of the site information. We already run into that a little bit with the system settings being outside of the database. 

Test case- say a live Fusion installation already exists on a production server, but a copy needs to be made on a local environment for development purposes. The developer would need to pull the Fusion database, but now also needs to pull the replicator files from the filesystem in order to proceed.

At the end of the day, option 2 is certainly a viable alternative to option 1, but not a solution to the problems raised by option 1.

**Comment by kaidesu** [comment]
Also of note, because we want to generate a table for each replicator fieldtype added to a fieldset, we still need that table to be **unique**. You could technically have a replicator with the same name on multiple fieldsets, but that doesn't mean they're all the same.

So when we generate tables for these, we should still generate a short random identifier to append to the table:

```
replicator_content_14jqa
```

## Validation Tip
The maximum character length for MySQL tables is 64 characters so `replicator_{handle}_14jhg` eats up `17` characters characters giving us a max length of `47` characters for replicator handles:

| Keyword | Length | Notes |
|---|---|---|
| `replicator` | `10` | |
| `__` | `2` | Underscore before and after handle. |
| `12abc` | `5` | Randomly generated 5 character alphanumeric string. |
